### PR TITLE
Re-implement adt/internedstateset via confluent persistent set

### PR DESCRIPTION
**Pull request by silentbicycle**
This replaces the interned state set ADT with a new implementation which should scale _much_ better, adapted from the Confluent Persistent Set described in _Confluently Persistent Sets and Maps_ by Olle Liljenzin.

The time and memory usage from the interned state set dominates the costs for `fsm_determinise`, which in turn dominate overall execution as the state machines become larger and more complex.

Casual benchmarking, using `tests/internedstateset/internedstateset_basic.c` with new parameter arguments, suggests this is dramatically more performant:

----
// calling `env LIMIT=$x SHUFFLE=0 /usr/bin/time build/tests/internedstateset/run_basic`

==== current implementation, 10 nodes
0.03user 0.00system 0:00.03elapsed 100%CPU (0avgtext+0avgdata 4992maxresident)k
0inputs+0outputs (0major+2786minor)pagefaults 0swaps

==== new implementation, 10 nodes
0.00user 0.00system 0:00.01elapsed 91%CPU (0avgtext+0avgdata 4916maxresident)k
0inputs+0outputs (0major+2766minor)pagefaults 0swaps

==== current implementation, 100 nodes
0.00user 0.00system 0:00.01elapsed 100%CPU (0avgtext+0avgdata 5732maxresident)k
0inputs+0outputs (0major+3167minor)pagefaults 0swaps

==== new implementation, 100 nodes
0.00user 0.00system 0:00.00elapsed 100%CPU (0avgtext+0avgdata 5148maxresident)k
0inputs+0outputs (0major+2876minor)pagefaults 0swaps

==== current implementation, 1000 nodes
0.09user 0.01system 0:00.10elapsed 100%CPU (0avgtext+0avgdata 26300maxresident)k
0inputs+0outputs (0major+10263minor)pagefaults 0swaps

==== new implementation, 1000 nodes
0.02user 0.00system 0:00.02elapsed 100%CPU (0avgtext+0avgdata 5624maxresident)k
0inputs+0outputs (0major+2988minor)pagefaults 0swaps

==== current implementation, 10000 nodes
7.45user 0.27system 0:07.72elapsed 99%CPU (0avgtext+0avgdata 1015560maxresident)k
0inputs+0outputs (0major+296829minor)pagefaults 0swaps

==== new implementation, 10000 nodes
0.21user 0.00system 0:00.21elapsed 100%CPU (0avgtext+0avgdata 8104maxresident)k
0inputs+0outputs (0major+3675minor)pagefaults 0swaps

==== current implementation, 100000 nodes
(does not terminate)

==== new implementation, 100000 nodes
2.46user 0.01system 0:02.48elapsed 99%CPU (0avgtext+0avgdata 35288maxresident)k
0inputs+0outputs (0major+11989minor)pagefaults 0swaps

----

While this is a somewhat contrived benchmark -- it's constructing [0..n], then again and checking that the root node ID matches, then [n..0] and checking the root node again -- testing with input data I can't publicly share also shows a huge improvement.

### CODE REVIEW

**Comment by silentbicycle** [comment]
It looks like all the CI failures were on macos.

**Comment by silentbicycle** [comment]
100000 nodes, 35 mb -> 25 mb
200000 nodes, 64 mb -> 44 mb

Yeah, I think the optimization in 1becb48 pays off. :grin:

File: src/adt/internedstateset.c [hunk 1]
```diff
@@ -15,436 +18,1226 @@
...
+			}
+			continue;
+		} else if (b_id == NO_ID) {
+			/* not found -- reserve node, save, and return */
+			iss_node_id new_id;
+			if (!freelist_pop(pool, &new_id)) {
+				return false;
+			}
+			iss.id = new_id;
+			memcpy(&pool->pool[new_id], &iss, sizeof(iss));
```
Comment (katef) [inline src/adt/internedstateset.c:678]: I think we can just use assignment here, can't we?
```
pool->pool[new_id] = iss;
```

File: src/adt/internedstateset.c [hunk 2]
```diff
@@ -15,436 +18,1226 @@
...
+					exit(EXIT_FAILURE);
+				}
+			}
+#endif
+
+			LOG_REF("new_node", new_id, "=1");
+			return true;
+		}
+
+		struct iss_node *b_iss = get_iss_node(pool, b_id);
```
Comment (katef) [inline src/adt/internedstateset.c:759]: can this be const?

Comment (silentbicycle) [reply src/adt/internedstateset.c]: Sure. Done in 848fa57a.

File: src/adt/internedstateset.c (hunk 1 above)
Comment (silentbicycle) [reply src/adt/internedstateset.c]: Done in 8b5b4447.

**Review by katef — APPROVED** [review]
:+1: pending CI, thank you very much!

**Comment by katef** [comment]
I removed CI builds for macos in https://github.com/katef/libfsm/pull/371

There's no advantage to merging from main to bring that in here; at the time of reviewing the only CI failures on this branch were for things which have now been removed. So I'm going to merge this PR as-is. Thank you

**Pull request merged**